kafka/src/main/java/org/axonframework/extensions/kafka/eventhandling/tokenstore/TokenStoreState.java

```diff
@@ -0,0 +1,310 @@
+            Map<String, String> topicConfig = new HashMap<>();
+            topicConfig.put(TopicConfig.CLEANUP_POLICY_CONFIG, TopicConfig.CLEANUP_POLICY_COMPACT);
+            //needed to make sure we have at least one new message in case of concurrent writes
+            topicConfig.put(
+                    TopicConfig.MIN_COMPACTION_LAG_MS_CONFIG,
+                    String.valueOf((2 * Duration.from(claimTimeout).getSeconds()))
+            );
+            topicConfig.put(
+                    TopicConfig.MAX_COMPACTION_LAG_MS_CONFIG,
+                    String.valueOf((8 * Duration.from(claimTimeout).getSeconds()))
+            );
+            newTopic.configs(topicConfig);
+            client.createTopics(Collections.singletonList(newTopic));
+        } catch (ExecutionException | InterruptedException e) {
+            logger.warn("error creating topic for Kafka token store", e);
+            Thread.currentThread().interrupt();
+        }
+    }
+
+    private void update(ConsumerRecord<String, TokenUpdate> consumerRecord) {
+        boolean result;
+        TokenUpdate update;
+        if (consumerRecord.value() != null) {
+            update = consumerRecord.value();
+            result = updateAddition(update);
+        } else {
+            update = new TokenUpdate(consumerRecord.headers(), null);
+            result = updateDeletion(update);
+        }
+        feedback.computeIfPresent(
+                update.getId(),
+                (k, f) -> {
+                    f.complete(result);
+                    return null;
+                });
+    }
+
+    private boolean updateAddition(TokenUpdate tokenUpdate) {
+        try {
+            updateState(tokenUpdate);
+            return true;
+        } catch (TokenStoreUpdateException e) {
+            logger.info("failed to update state for '{}[{}]'",
+                        tokenUpdate.getProcessorName(),
+                        tokenUpdate.getSegment(),
+                        e);
+            return false;
+        }
+    }
+
+    private boolean updateDeletion(TokenUpdate tokenUpdate) {
+        final AtomicBoolean result = new AtomicBoolean(false);
+        state.computeIfPresent(tokenUpdate.getProcessorName(), (k, v) -> {
+            result.set(v.remove(tokenUpdate.getSegment()) != null);
+            return v;
+        });
+        return result.get();
+    }
+
+    private void updateState(TokenUpdate tokenUpdate) {
+        Map<Integer, TokenUpdate> innerMap = state.computeIfAbsent(
+                tokenUpdate.getProcessorName(),
+                x -> new ConcurrentHashMap<>());
+        innerMap.merge(tokenUpdate.getSegment(), tokenUpdate, this::mergeFunction);
+    }
+
+    private TokenUpdate mergeFunction(TokenUpdate oldUpdate, TokenUpdate newUpdate) {
+        if (newUpdate.getSequenceNumber() > oldUpdate.getSequenceNumber()) {
+            return newUpdate;
+        } else {
+            throw new TokenStoreUpdateException(String.format(SEQUENCE_ERROR_FORMAT,
```

Comment:
Curious: could you describe a scenario when this might happen? And, what follow-up steps a user would need to take to recover from it, if any?

Reply:
It can only happen when there are two updates with the same sequence number short after each other. Since it's only needed in the `TokenStorState` I made it a private static class instead.